File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {

    public void soln(String digits, ArrayList<String> keypad, String curr, ArrayList<String> result)
    {
        if(digits.isEmpty())
        {
            if(!curr.isEmpty())
                result.add(curr);
            return;
        }

        int dig = (int)digits.charAt(0)-(int)'0';
        for(int i=0; i<keypad.get(dig).length(); i++)
        {
            soln(digits.substring(1, digits.length()), keypad, curr+keypad.get(dig).charAt(i), result); 
        }
    }

    public List<String> letterCombinations(String digits) 
    {
        // String array for the keypad
        ArrayList<String> keypad = new ArrayList<>(Arrays.asList("", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"));

        // String array to store result
        ArrayList<String> result = new ArrayList<>();
        soln(digits, keypad, new String(), result);
        return result;   
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void helper(string digits, string temp, unordered_map<char, string> &map, vector<string> &ans)
{
    if(digits == ""){
        if(!temp.empty())
        ans.push_back(temp);
        return;
    }

    string digs = map[digits[0]];

    for(int i=0; i<digs.size(); i++)
    {
        temp.push_back(digs[i]);
        helper(digits.substr(1), temp, map, ans);
        temp.pop_back();
    }

}

vector<string> letterCombinations(string digits) {
    unordered_map<char, string> map;
    map['2'] = "abc";
    map['3'] = "def";
    map['4'] = "ghi";
    map['5'] = "jkl";
    map['6'] = "mno";
    map['7'] = "pqrs";
    map['8'] = "tuv";
    map['9'] = "wxyz";

    vector<string> ans;
    helper(digits, "", map, ans);
    return ans;
}
};